Improve transaction relay logic

ximinez opened the pull request:
## High Level Overview of Change

This PR, if merged, will improve transaction relay logic around a few edge cases.

(I'll write a single commit message later, but before this PR is squashed and merged.)

### Context of Change

A few months ago, while examining some of the issues around the 2.0.0 release, and auditing transaction relay code, I identified a few areas with potential for improvement. 

### Type of Change

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Tests (you added tests for code that already exists, or your new feature included in this PR)

## Before / After

This PR is divided into four mostly independent changes.
1. "Decrease `shouldRelay` limit to 30s." Pretty self-explanatory. Currently, the limit is 5 minutes, by which point the `HashRouter` entry could have expired, making this transaction look brand new (and thus causing it to be relayed back to peers which have sent it to us recently).
2.  "Give a transaction more chances to be retried." Will put a transaction into `LedgerMaster`'s held transactions if the transaction gets a `ter`, `tel`, or `tef` result. Old behavior was just `ter`.
     * Additionally, to prevent a transaction from being repeatedly held indefinitely, it must meet some extra conditions. (Documented in a comment in the code.)
3. "Pop all transactions with sequential sequences, or tickets." When a transaction is processed successfully, currently, one held transaction for the same account (if any) will be popped out of the held transactions list, and queued up for the next transaction batch. This change pops all transactions for the account, but only if they have sequential sequences (for non-ticket transactions) or use a ticket. This issue was identified from interactions with @mtrippled's #4504, which was merged, but unfortunately reverted later by #4852. When the batches were spaced out, it could potentially take a very long time for a large number of held transactions for an account to get processed through. However, whether batched or not, this change will help get held transactions cleared out, particularly if a missing earlier transaction is what held them up.
4. "Process held transactions through existing NetworkOPs batching." In the current processing, at the end of each consensus round, all held transactions are directly applied to the open ledger, then the held list is reset. This bypasses all of the logic in `NetworkOPs::apply` which, among other things, broadcasts successful transactions to peers. This means that the transaction may not get broadcast to peers for a really long time (5 minutes in the current implementation, or 30 seconds with this first commit). If the node is a bottleneck (either due to network configuration, or because the transaction was submitted locally), the transaction may not be seen by any other nodes or validators before it expires or causes other problems.

ximinez committed:
Decrease `shouldRelay` limit to 30s:

* Allows transactions, validator lists, proposals, and validations to be
  relayed more often, but only when triggered by another event, such as
  receiving it from a peer
* Decrease from 5min.
* Expected to help transaction throughput on poorly connected networks.

ximinez committed:
Give a transaction more chances to be retried:

* Hold if the transaction gets a ter, tel, or tef result.
* Use the new SF_HELD flag to ultimately prevent the transaction from
  being held and retried too many times.

ximinez committed:
Pop all transactions with sequential sequences, or tickets

ximinez committed:
Process held transactions through existing NetworkOPs batching:

* Ensures that successful transactions are broadcast to peers,
  appropriate failed transactions are held for later attempts, fee
  changes are sent to subscribers, etc.

scottschurr commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -1298,7 +1333,59 @@ NetworkOPsImp::doTransactionSync(
+        mTransactions.swap(transactions);
+    else
+    {
+        for (auto& t : transactions)
```
Does it make sense to reserve space in `mTransactions`?  Consider
```
mTransactions.reserve(mTransactions.size() + transactions.size());
```

ximinez replied:
Yes it does!

scottschurr reviewed (commented):
Very nice!  I really appreciate the way the commits were divided up.  It made the code review much easier.

I mostly left complementary comments.  But there's one `bool` that I suspect needs to be changed to a `std::atomic<bool>`.  See what you think...

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  fix: Remove redundant STAmount conversion in test (4996)
  fix: resolve database deadlock: (4989)
  test: verify the rounding behavior of equal-asset AMM deposits (4982)
  test: Add tests to raise coverage of AMM (4971)
  chore: Improve codecov coverage reporting (4977)
  test: Unit test for AMM offer overflow (4986)
  fix amendment to add `PreviousTxnID`/`PreviousTxnLgrSequence` (4751)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.2.0-b3

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Ignore more commits
  Address compiler warnings
  Add markers around source lists
  Fix source lists
  Rewrite includes
  Format formerly .hpp files
  Rename .hpp to .h
  Simplify protobuf generation
  Consolidate external libraries
  Remove packaging scripts
  Remove unused files

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.2.0-rc1

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Remove flow assert: (5009)
  Update list of maintainers: (4984)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Add external directory to Conan recipe's exports (5006)
  Add missing includes (5011)

ximinez committed:
Merge commit 'c706926' into relay

* commit 'c706926': (23 commits)
  Change order of checks in amm_info: (4924)
  Add the fixEnforceNFTokenTrustline amendment: (4946)
  Replaces the usage of boost::string_view with std::string_view (4509)
  docs: explain how to find a clang-format patch generated by CI (4521)
  XLS-52d: NFTokenMintOffer (4845)
  chore: remove repeat words (5041)
  Expose all amendments known by libxrpl (5026)
  fixReducedOffersV2: prevent offers from blocking order books: (5032)
  Additional unit tests for testing deletion of trust lines (4886)
  Fix conan typo: (5044)
  Add new command line option to make replaying transactions easier: (5027)
  Fix compatibility with Conan 2.x: (5001)
  Set version to 2.2.0
  Set version to 2.2.0-rc3
  Add xrpl.libpp as an exported lib in conan (5022)
  Fix Oracle's token pair deterministic order: (5021)
  Set version to 2.2.0-rc2
  Fix last Liquidity Provider withdrawal:
  Fix offer crossing via single path AMM with transfer fee:
  Fix adjustAmountsByLPTokens():
  ...

ximinez committed:
Merge commit 'f6879da' into relay

* commit 'f6879da':
  Add bin/physical.sh (4997)
  Prepare to rearrange sources: (4997)

Commit:
Move CMake directory

Commit:
Rearrange sources

Commit:
Rewrite includes

Commit:
Recompute loops

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  fixInnerObjTemplate2 amendment (5047)
  Set version to 2.3.0-b1
  Ignore restructuring commits (4997)
  Recompute loops (4997)
  Rewrite includes (4997)
  Rearrange sources (4997)
  Move CMake directory (4997)

ximinez committed:
Fix formatting

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  fix CTID in tx command returns invalidParams on lowercase hex (5049)
  Invariant: prevent a deleted account from leaving (most) artifacts on the ledger. (4663)
  Bump codecov plugin version to version 4.5.0 (5055)
  fix "account_nfts" with unassociated marker returning issue (5045)

intelliot commented:
would you like to open the request to have @sophiax851 take a look at this?

ximinez commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -1401,15 +1488,20 @@ NetworkOPsImp::apply(std::unique_lock<std::mutex>& batchLock)
-                auto const txNext = m_ledgerMaster.popAcctTransaction(txCur);
-                if (txNext)
+                auto txNext = m_ledgerMaster.popAcctTransaction(txCur);
+                while (txNext)
```
To clarify, this loop doesn't do anything with the `TxQ`. This is reading the list of held transactions in `LedgerMaster`. So, AFAIK, there is no limit. However, the only way for a tx to get into that list is in this same function (see calls to `addHeldTransaction`), and only transactions received directly from RPC or from peers get to this function. That said, a malicious client or peer could probably get a bunch of transactions held (this is true without these changes). Thus, it might make sense to add a limit to mitigate some kinds of attacks.

Also, `submit_held` is a `std::vector`. Anything added to it will be processed in the _next_ batch of transactions.

ximinez commented on src/ripple/app/ledger/impl/LedgerMaster.cpp:
```diff
@@ -548,26 +548,17 @@ LedgerMaster::storeLedger(std::shared_ptr<Ledger const> ledger)
+        // VFALCO NOTE The hash for an open ledger is undefined so we use
+        // something that is a reasonable substitute.
+        CanonicalTXSet set(app_.openLedger().current()->info().parentHash);
+        std::swap(mHeldTransactions, set);
```
Thanks!

ximinez commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -1187,14 +1200,9 @@ NetworkOPsImp::submitTransaction(std::shared_ptr<STTx const> const& iTrans)
-    bool bLocal,
-    FailHard failType)
+bool
+NetworkOPsImp::preProcessTransaction(std::shared_ptr<Transaction>& transaction)
```
Thanks!

ximinez commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -1276,6 +1300,17 @@ NetworkOPsImp::doTransactionSync(
 
+    doTransactionSyncBatch(
+        lock, [&transaction](std::unique_lock<std::mutex>& lock) {
+            return transaction->getApplying();
```
I think this was done intentionally because it avoids the overhead of taking a lock, and the consequences of a race condition are nearly-zero.

1. Most significantly, almost all accesses of the `*Applying` functions are under `NetworkOPsImp`'s lock. The exceptions are
    1. The new `processTransactionSet` function.
    2. The `popAcctTransaction` loop we were just talking about.
3. If there is a race condition, and `getApplying` returns false when it should be true, the transaction will be processed again. Not that big a deal if it's a rare one-off. Most of the time, it'll get `tefALREADY` or `tefPAST_SEQ`.
4. On the flip side, if it returns true, when it should be false, then the transaction must have been attempted recently, so no big deal if it doesn't immediately get tried right away.
5. If there's a race between `setApplying` and `clearApplying`, again, the flag ends up set, then a batch is about to try to process the transaction and will call `clearApplying` later. If it ends up cleared, then it might get attempted again later as is the case with item 2.

Alllll that said, I don't think it would hurt to rewrite those two lock exceptions I pointed out to ensure that the lock is taken for the function calls.

ximinez commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -1224,12 +1232,28 @@ NetworkOPsImp::processTransaction(
         transaction->setResult(temBAD_SIGNATURE);
         app_.getHashRouter().setFlags(transaction->getID(), SF_BAD);
-        return;
+        return false;
```
It should be impossible to get to this line. Note the block above explains that I don't think the check is necessary, and has an `assert` on the same `validity` at the end.
```
    // NOTE eahennis - I think this check is redundant,
    // but I'm not 100% sure yet.
    // If so, only cost is looking up HashRouter flags.
    auto const view = m_ledgerMaster.getCurrentLedger();
    auto const [validity, reason] = checkValidity(
        app_.getHashRouter(),
        *transaction->getSTransaction(),
        view->rules(),
        app_.config());
    assert(validity == Validity::Valid);
```

ximinez commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -1203,7 +1211,7 @@ NetworkOPsImp::processTransaction(
         transaction->setStatus(INVALID);
         transaction->setResult(temBAD_SIGNATURE);
-        return;
+        return false;
```
Yeah, this one isn't documented as explicitly, but the idea is that it should also be impossible to get to this point with a bad signature.

ximinez commented:
> would you like to open the request to have @sophiax851 take a look at this?

Done: https://ripplelabs.atlassian.net/browse/RPFC-173

ximinez committed:
[FOLD] Review feedback from @scottschurr:

* Limit number of held transactions popped for retry after a successful
  transaction.
* Check and set the Transaction::mApplying flag under NetworkOPsImp's
  lock.
* Reserve space when appending to transaction vector(s)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  chore: remove repeat words (5053)

sophiax851 commented:
@ximinez I'm reading the changes in this PR and trying to think if our release payment load testing case would come across the processing path modified by this PR. Based on the highlevel overview:
1. The 5mins to 30sec change. I think this change is definitely valid, but the transactions in our test case wouldn't be held that long unless something configured very wrong
2. "Give a transaction more chances to be retried." This one won't be invoked either in our load, hopefully we have integration test case to cover the testing in functionality
3. "Pop all transactions with sequential sequences, or tickets." To properly test this, we need to somehow simulate a case where a lot of transactions were held for the same account but somehow there's missing sequence(s) among them. This would never happen if our CH node is doing its job, right? Maybe I'm missing something?
4. "Process held transactions through existing NetworkOPs batching." I think the only way that we could possibly test this one is to send transactions load directly to the validators? In normal case, transactions being reapplied in the validators were previously received via broadcasting from other nodes, right? Would we add additional load to the network if we broadcast transactions again?

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Add xrpld build option and Conan package test (5052)

ximinez committed:
fixup! [FOLD] Review feedback from @scottschurr:

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Update BUILD.md after PR 5052 (5067)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  chore: Add comments to SignerEntries.h (5059)
  chore: Rename two files from Directory* to Dir*: (5058)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Ensure levelization sorting is ASCII-order across platforms (5072)
  fix: Fix NuDB build error via Conan patch (5061)
  Disallow filtering account_objects by unsupported types (5056)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Update gcovr EXCLUDE (5084)
  Fix crash inside `OverlayImpl` loops over `ids_` (5071)
  Set version to 2.3.0-b2
  docs: Document the process for merging pull requests (5010)
  Remove unused constants from resource/Fees.h (4856)
  fix: change error for invalid `feature` param in `feature` RPC (5063)
  Set version to 2.2.1
  Use error codes throughout fast Base58 implementation
  Improve error handling in some RPC commands

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Factor out Transactor::trapTransaction (5087)
  Remove shards (5066)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Address rare corruption of NFTokenPage linked list (4945)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  chore: Fix documentation generation job: (5091)
  chore: libxrpl verification on CI (5028)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  docs: Update options documentation (5083)
  refactor: Remove dead headers (5081)
  refactor: Remove reporting mode (5092)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Update Release Notes for 2.2.1 and 2.2.2
  Set version to 2.2.2
  Allow only 1 job queue slot for each validation ledger check
  Allow only 1 job queue slot for acquiring inbound ledger.
  Track latencies of certain code blocks, and log if they take too long

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  test: Retry RPC commands to try to fix MacOS CI jobs (5120)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.3.0-b4
  feat(SQLite): allow configurable database pragma values (5135)
  refactor: re-order PRAGMA statements (5140)
  fix(book_changes): add "validated" field and reduce RPC latency (5096)
  chore: fix typos in comments (5094)
  Set version to 2.2.3
  Update SQLite3 max_page_count to match current defaults (5114)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Expand Error Message for rpcInternal (4959)
  docs: clean up API-CHANGELOG.md (5064)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Consolidate definitions of fields, objects, transactions, and features (5122)
  Ignore reformat when blaming
  Reformat code with clang-format-18
  Update pre-commit hook
  Update clang-format settings
  Update clang-format workflow

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  docs: Add protobuf dependencies to linux setup instructions (5156)
  fix: reject invalid markers in account_objects RPC calls (5046)
  Update RELEASENOTES.md (5154)
  Introduce MPT support (XLS-33d): (5143)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Add hubs.xrpkuwait.com to bootstrap (5169)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Add AMMClawback Transaction (XLS-0073d) (5142)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Fix unity build (5179)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.3.0-rc1
  Replace Uint192 with Hash192 in server_definitions response (5177)
  Fix potential deadlock (5124)
  Introduce Credentials support (XLS-70d): (5103)
  Fix token comparison in Payment (5172)
  Add fixAMMv1_2 amendment (5176)

vlntb commented on src/xrpld/app/misc/NetworkOPs.cpp:
```diff
@@ -1396,16 +1493,28 @@ NetworkOPsImp::apply(std::unique_lock<std::mutex>& batchLock)
+
+                std::size_t count = 0;
+                for (auto txNext = m_ledgerMaster.popAcctTransaction(txCur);
+                     txNext && count < maxPoppedTransactions;
```
If we only pop a maximum of ten transactions, does this mean that over time, we will have a slow memory leak here of transactions that will never be used but will stay in the CanonicalTXSet indefinitely?

ximinez replied:
No, because the `mHeldTransactions` list that this function pulls from is also fully cleared out and retried all at once in `LedgerMaster::applyHeldTransactions()`. `applyHeldTransactions` is called during consensus when a new closed ledger is created from the open ledger to start the process.

There are restrictions on how many times a transaction can be put into `mHeldTransactions` so that it can't cycle in and out indefinitely. Those are described at one of the call sites of `addHeldTransaction()`

vlntb reviewed (commented):
The change was thoroughly reviewed by former colleagues. I added a couple of questions. Once all the questions are answered, I will be happy to approve it.

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  fix: include `index` in `server_definitions` RPC (5190)
  Fix ledger_entry crash on invalid credentials request (5189)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.3.0-rc2

ximinez committed:
Add UNDOCUMENTED configuration section for HashRouter timeouts

ximinez commented on src/xrpld/app/misc/HashRouter.h:
```diff
@@ -221,6 +235,7 @@ class HashRouter
         suppressionMap_;
 
     std::chrono::seconds const holdTime_;
+    std::chrono::seconds const relayTime_;
```
The `HoldTime` has been 300s for [over 11 years](https://github.com/ximinez/rippled/commit/699ac3d25669f04f3b30974100df16d1cb287499), so it's a pretty safe bet that's a good value. However, I see your point about being able to tweak `RelayTime`.

I added a commit to add configuration options for these two values. I'm leaving them *undocumented* for now, because I don't want operators just randomly changing them for no particular reason.

ximinez commented on src/xrpld/app/misc/NetworkOPs.cpp:
```diff
@@ -1271,6 +1296,17 @@ NetworkOPsImp::doTransactionSync(
     }
 
+    doTransactionSyncBatch(
+        lock, [&transaction](std::unique_lock<std::mutex>& lock) {
```
Yeah, good idea.  I also changed the function signature to require the lock to be `const` in the callback.

ximinez committed:
Document why a lock is not used for Transaction::mApplying

vlntb reviewed (approved):
All raised questions addressed.

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.3.0
  refactor(AMMClawback): move tfClawTwoAssets check (5201)
  Add a new serialized type: STNumber (5121)
  fix: check for valid ammID field in amm_info RPC (5188)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  test: Add more test cases for Base58 parser (5174)
  test: Check for some unlikely null dereferences in tests (5004)
  Add Antithesis intrumentation (5042)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  refactor: clean up `LedgerEntry.cpp` (5199)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Enforce levelization in libxrpl with CMake (5111)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.4.0-b1
  fix: Add header for set_difference (5197)
  fix: allow overlapping types in `Expected` (5218)
  Add MPTIssue to STIssue (5200)
  Antithesis instrumentation improvements (5213)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  chore: add macos dependency installation (5233)
  prefix Uint384 and Uint512 with Hash in server_definitions (5231)
  refactor: add `rpcName` to `LEDGER_ENTRY` macro (5202)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  chore: update deprecated Github Actions (5241)
  XLS-46: DynamicNFT (5048)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  chore: Update Visual Studio CI to VS 2022, and add VS Debug builds (5240)
  Add [validator_list_threshold] to validators.txt to improve UNL security (5112)
  Switch from assert to XRPL_ASSERT (5245)
  Add missing space character to a log message (5251)
  Cleanup API-CHANGELOG.md (5207)
  test: Unit tests to recreate invalid index logic error (5242)
  Update branch management and merge / release processes  (5215)
  fix: Error consistency in LedgerEntry::parsePermissionedDomains() (5252)
  Set version to 2.4.0-b2
  fix: Use consistent CMake settings for all modules (5228)
  Fix levelization script to ignore commented includes (5194)
  Fix the flag processing of NFTokenModify (5246)
  Fix failing assert in `connect` RPC (5235)
  Permissioned Domains (XLS-80d) (5161)

ximinez committed:
fixup! Add UNDOCUMENTED configuration section for HashRouter timeouts

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Fix CI unit tests (5196)
  Update secp256k1 library to 0.6.0 (5254)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.4.0-b3
  Set version to 2.3.1
  Update conan in the "nix" CI jobs
  Add RPC "simulate" to execute a dry run of a transaction (5069)
  Set version to 2.3.1-rc1
  Reduce the peer charges for well-behaved peers:

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Add deep freeze feature (XLS-77d) (5187)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Amendment `fixFrozenLPTokenTransfer` (5227)
  Improve git commit hash lookup (5225)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Updates Conan dependencies (5256)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  fix: Do not allow creating Permissioned Domains if credentials are not enabled (5275)
  fix: issues in `simulate` RPC (5265)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  fix: Amendment to add transaction flag checking functionality for Credentials (5250)
  fix: Omit superfluous setCurrentThreadName call in GRPCServer.cpp (5280)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  chore: Rename missing-commits job, and combine nix job files (5268)
  docs: Add a summary of the git commit message rules (5283)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  fix: Replace charge() by fee_.update() in OnMessage functions (5269)
  docs: ensure build_type and CMAKE_BUILD_TYPE match (5274)
  chore: Fix small typos in protocol files (5279)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Reduce duplicate peer traffic for ledger data (5126)
  Support canonical ledger entry names (5271)
  chore: Update XRPL Foundation public key (5289)
  refactor: Change recursive_mutex to mutex in DatabaseRotatingImp (5276)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Revert "Reduce duplicate peer traffic for ledger data (5126)" (5300)
  docs: Revert peer port to 51235 (5299)
  Set version to 2.4.0-rc1
  fix: Switch Permissioned Domain to Supported::yes (5287)
  docs: Clarifies default port of hosts (5290)
  Log proposals and validations (5291)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Add logging and improve counting of amendment votes from UNL (5173)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.4.0-rc2

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  fix: Remove 'new parent hash' assert (5313)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.4.0-rc3
  fix: Acquire previously failed transaction set from network as new proposal arrives (5318)
  Fix Replace `assert` with `XRPL_ASSERT` (5312)

Bronek commented on src/xrpld/app/misc/NetworkOPs.cpp:
```diff
@@ -1298,7 +1334,68 @@ NetworkOPsImp::doTransactionSync(
+    }
+
+    doTransactionSyncBatch(lock, [&](std::unique_lock<std::mutex> const&) {
+        assert(lock.owns_lock());
```
please use `XRPL_ASSERT`

ximinez replied:
> please use `XRPL_ASSERT`

Good catch. Thanks.

Bronek reviewed (commented):
Partial review

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Log detailed correlated consensus data together (5302)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  chore: Move "assert" and "werr" flags from "actions/build" (5325)

ximinez commented on src/xrpld/app/misc/HashRouter.cpp:
```diff
@@ -122,10 +124,45 @@ HashRouter::shouldRelay(uint256 const& key)
+
+    if (set(tmp, "hold_time", section))
+    {
+        if (tmp < 8)
```
> The error message below says something different; is this meant to be 8 or 12 ?

This looks like a copy/paste error. Unfortunately, I don't remember which direction. LOL.

ximinez commented on src/xrpld/app/misc/HashRouter.h:
```diff
@@ -56,6 +59,27 @@ class HashRouter
+    /** Structure used to customize @ref HashRouter behavior.
+     *
+     * Even though these items are configurable, don't change them unless there
+     * is a good reason, and network-wide coordination to do it.
```
This is a pretty common pattern throughout `rippled`. For example `FeeSetup` and `setup_FeeVote`. The difference is that `FeeSetup` is documented. It's done this way with default values so that the structures can be used in, for example, unit tests using either the default values, _or_ custom values without needing the overhead of having to populate a `Config` object. It's also handy to be able to look in one place to see what the defaults are. Most of the objects that use this pattern are singletons that are created on startup, so it's not like anybody is going to forget to call the `setup_` function.

This isn't the only undocumented configuration feature in `rippled`. Those exist intentionally, because in some situations, it can be a lot faster to communicate with operators (especially UNL operators) and get them to update their config than it would be to roll out a new version, even if it's a hotfix.

Based on your feedback, I did update this comment to reflect
1. The fact that it's undocumented.
2. That `setup_HashRouter` processes the config.

Bronek replied:
Thank you, that works

ximinez commented on src/xrpld/app/misc/HashRouter.cpp:
```diff
@@ -122,10 +124,45 @@ HashRouter::shouldRelay(uint256 const& key)
     auto& s = emplace(key).first;
 
-    if (!s.shouldRelay(suppressionMap_.clock().now(), holdTime_))
+    if (!s.shouldRelay(suppressionMap_.clock().now(), setup_.relayTime))
```
> Can we have a comment here why `relayTime` value is used for `holdTime` parameter ? Also, it appears to be the only use of the `Entry::shouldRelay` function, perhaps this parameter needs to be renamed to `relayTime` or `retryTime` ?

The parameter in `s.shouldRelay` is named `holdTime` because that's the value that was used before I introduced the shorter relay retry cutoff. I'll rename it.

ximinez committed:
Address review feedback from @Bronek:

- Add a HashRouter test case for setup_HashRouter config processing.
- Correct a typo in setup_HashRouter.
- Add a missing include.
- Update the documentation for HashRouter::Setup.
- Update the parameter name of HashRouter::Entry::shouldRelay.

ximinez committed:
Add missing include

ximinez committed:
Get rid of some unused variables

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  Set version to 2.4.0
  Set version to 2.4.0-rc4
  chore: Update XRPL Foundation Validator List URL (5326)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  refactor: Remove unused and add missing includes (5293)

ximinez committed:
Merge remote-tracking branch 'upstream/develop' into relay

* upstream/develop:
  refactor: Improve ordering of headers with clang-format (5343)
  refactor: Calculate numFeatures automatically (5324)

ximinez committed:
Fix formatting

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

ximinez committed:
Merge branch 'develop' into relay

Bronek commented on src/xrpld/app/misc/CanonicalTXSet.cpp:
```diff
@@ -66,18 +66,22 @@ CanonicalTXSet::popAcctTransaction(std::shared_ptr<STTx const> const& tx)
-        itrNext->first.getAccount() == effectiveAccount)
+        itrNext->first.getAccount() == effectiveAccount &&
+        (!itrNext->second->getSeqProxy().isSeq() ||
+         itrNext->second->getSeqProxy().value() == seqProxy.value() + 1))
```
could this produce different transaction order compared to the old code ? If so, shouldn't this be amendment-gated ?

ximinez replied:
> could this produce different transaction ~order~ set compared to the old code ? If so, shouldn't this be amendment-gated ?

No, because this transaction is only relevant in the context of `NetworkOPsImp::apply`, which only handles the open ledger. Ordering rules are enforced, but don't matter the way they do in building a consensus ledger. Also, the one use is in `LedgerMaster::popAccountTransaction`, and is done on `mHeldTransactions`, which are transactions that are held for retry into the open ledger. Again, nothing to do with consensus.

Bronek commented on src/xrpld/app/misc/NetworkOPs.cpp:
```diff
@@ -1441,16 +1540,28 @@ NetworkOPsImp::apply(std::unique_lock<std::mutex>& batchLock)
                     t->setApplying();
                 }
+                if (batchLock.owns_lock())
+                    batchLock.unlock();
```
I don't love that this code **needs** to do it, but I agree that's the best (most robust) way to deal with the problem. Approved.

ximinez replied:
> I don't love that this code **needs** to do it, but I agree that's the best (most robust) way to deal with the problem. Approved.

Yeah, it's a compromise between not wanting to take the lock outside the loop, which would lock every time a transaction succeeds, not knowing if there are any transactions to pop, and not wanting to take the lock inside the loop, which would churn a lot for what should be a very quick loop.

Bronek commented on src/xrpld/app/misc/NetworkOPs.cpp:
```diff
@@ -1472,16 +1583,54 @@ NetworkOPsImp::apply(std::unique_lock<std::mutex>& batchLock)
                 e.transaction->setKept();
             }
-            else if (isTerRetry(e.result))
+            else if (
```
again, this I think could cause a difference in transactions sets, making me wonder how safe is this change without an amendment.

ximinez replied:
> again, this I think could cause a difference in transactions sets, making me wonder how safe is this change without an amendment.

Again, this is in the context of the open ledger, which won't affect consensus or the validated ledger.

Bronek reviewed (approved):
All the changes here are an improvement. There's one open question "does this need amendment ?" and it can be most likely addressed by a PR comment here (assuming the answer is "no", which I feel is likely).

Bronek committed:
Merge branch 'develop' into relay

ximinez commented:
> All the changes here are an improvement. There's one open question "does this need amendment ?" and it can be most likely addressed by a PR comment here (assuming the answer is "no", which I feel is likely).

No. 😀 An amendment is not needed for these changes. I am running some "due diligence" on my local node right now to make sure there are no obvious negative side effects like de-syncs.

Bronek committed:
Merge branch 'develop' into relay
